From: tadam <tadam>
Date: Tue, 22 Jan 2013 08:32:34 +0000
Subject: [PATCH] Comply with Debian's hardening rules.

Debian's policy of what's considered "fit for purpos" for inclusion in
its repositoy of packages says that all compiled programs must pass the
following:

CFLAGS="-I/usr/local/include -g -O2 -fstack-protector \
		--param=ssp-buffer-size=4 -Wformat -Werror=format-security" \
LDFLAGS="-L/usr/local/lib"

http://wiki.debian.org/HardeningWalkthrough

Conflicts:
	ChangeLog
	modules/ChangeLog

diff --git a/fvwm/builtins.c b/fvwm/builtins.c
index 8cb39e5..301c6e9 100644
--- a/fvwm/builtins.c
+++ b/fvwm/builtins.c
@@ -2682,7 +2682,7 @@ void CMD_Echo(F_CMD_ARGS)
 			action[len-1]='\0';
 		}
 	}
-	fvwm_msg(ECHO,"Echo",action);
+	fvwm_msg(ECHO, "", "Echo %s", action);
 
 	return;
 }
diff --git a/fvwm/misc.c b/fvwm/misc.c
index f303453..3c57cc0 100644
--- a/fvwm/misc.c
+++ b/fvwm/misc.c
@@ -346,6 +346,7 @@ static char *fvwm_msg_strings[] =
 void fvwm_msg(fvwm_msg_t type, char *id, char *msg, ...)
 {
 	va_list args;
+	char *mfmt;
 	char fvwm_id[20];
 	char time_str[40] = "\0";
 #ifdef FVWM_DEBUG_TIME
@@ -378,21 +379,22 @@ void fvwm_msg(fvwm_msg_t type, char *id, char *msg, ...)
 		sprintf(&fvwm_id[strlen(fvwm_id)], ".%d", (int)Scr.screen);
 	}
 
-	fprintf(stderr, "%s[%s][%s]: %s",
+	fprintf(stderr, "%s[%s][%s]: %s\n",
 		time_str, fvwm_id, id, fvwm_msg_strings[(int)type]);
 
 	if (type == ECHO)
 	{
 		/* user echos must be printed as a literal string */
-		fprintf(stderr, "%s", msg);
+		fprintf(stderr, "%s\n", msg);
 	}
 	else
 	{
 		va_start(args, msg);
-		vfprintf(stderr, msg, args);
+		asprintf(&mfmt, "%s\n", msg);
+		vfprintf(stderr, mfmt, args);
 		va_end(args);
+		free(mfmt);
 	}
-	fprintf(stderr, "\n");
 
 	if (type == ERR)
 	{
@@ -411,6 +413,7 @@ void fvwm_msg(fvwm_msg_t type, char *id, char *msg, ...)
 		}
 		BroadcastName(M_ERROR, 0, 0, 0, tmp);
 	}
+	fprintf(stderr, "\n");
 
 } /* fvwm_msg */
 #endif
diff --git a/fvwm/move_resize.c b/fvwm/move_resize.c
index 22d16d0..941356d 100644
--- a/fvwm/move_resize.c
+++ b/fvwm/move_resize.c
@@ -3037,8 +3037,7 @@ void CMD_SnapAttraction(F_CMD_ARGS)
 	fvwm_msg(
 		OLD, "CMD_SnapAttraction",
 		"The command SnapAttraction is obsolete. Please use the"
-		" following command instead:");
-	fvwm_msg(OLD, "", cmd);
+		" following command instead:\n\n%s", cmd);
 	execute_function(
 		cond_rc, exc, cmd,
 		FUNC_DONT_REPEAT | FUNC_DONT_EXPAND_COMMAND);
@@ -3059,8 +3058,7 @@ void CMD_SnapGrid(F_CMD_ARGS)
 	fvwm_msg(
 		OLD, "CMD_SnapGrid",
 		"The command SnapGrid is obsolete. Please use the following"
-		" command instead:");
-	fvwm_msg(OLD, "", cmd);
+		" command instead:\n\n%s", cmd);
 	execute_function(
 		cond_rc, exc, cmd,
 		FUNC_DONT_REPEAT | FUNC_DONT_EXPAND_COMMAND);
diff --git a/fvwm/virtual.c b/fvwm/virtual.c
index 7f7cadf..80abe93 100644
--- a/fvwm/virtual.c
+++ b/fvwm/virtual.c
@@ -2019,10 +2019,7 @@ void CMD_EdgeResistance(F_CMD_ARGS)
 			OLD, "CMD_EdgeResistance",
 			"The command EdgeResistance with three arguments is"
 			" obsolete. Please use the following commands"
-			" instead:");
-		fvwm_msg(OLD, "", cmd);
-		fvwm_msg(OLD, "", stylecmd);
-		fvwm_msg(OLD, "", stylecmd2);
+			" instead:\n%s\n%s\n%s\n", cmd, stylecmd, stylecmd2);
 		execute_function(
 			cond_rc, exc, cmd,
 			FUNC_DONT_REPEAT | FUNC_DONT_EXPAND_COMMAND);
diff --git a/libs/FRenderInit.c b/libs/FRenderInit.c
index fcdb259..6dc8f36 100644
--- a/libs/FRenderInit.c
+++ b/libs/FRenderInit.c
@@ -129,7 +129,7 @@ Bool FRenderGetErrorText(int code, char *msg)
 		    code <= FRenderErrorBase +
 		    (sizeof(error_names) / sizeof(char *)) -1)
 		{
-			sprintf(msg, error_names[code - FRenderErrorBase]);
+			sprintf(msg, "%s\n", error_names[code - FRenderErrorBase]);
 			return 1;
 		}
 	}
diff --git a/modules/FvwmIconMan/debug.c b/modules/FvwmIconMan/debug.c
index 9b5c194..3aef888 100644
--- a/modules/FvwmIconMan/debug.c
+++ b/modules/FvwmIconMan/debug.c
@@ -37,6 +37,7 @@ int WINLIST = 0;
 void
 ConsoleMessage(const char *fmt, ...)
 {
+	char *mfmt;
 	va_list args;
 
 	assert(console != NULL);
@@ -44,8 +45,10 @@ ConsoleMessage(const char *fmt, ...)
 	fputs("FvwmIconMan: ", console);
 
 	va_start(args, fmt);
-	vfprintf(console, fmt, args);
+	asprintf(&mfmt, "%s\n", fmt);
+	vfprintf(console, mfmt, args);
 	va_end(args);
+	free(mfmt);
 }
 
 int
diff --git a/modules/FvwmIconMan/functions.c b/modules/FvwmIconMan/functions.c
index 77f3910..f841de9 100644
--- a/modules/FvwmIconMan/functions.c
+++ b/modules/FvwmIconMan/functions.c
@@ -534,8 +534,7 @@ int builtin_print(int numargs, BuiltinArg *args)
 	}
 	else
 	{
-		sprintf(buf, "%s\n", s);
-		ConsoleMessage(buf);
+		ConsoleMessage("%s\n", s);
 	}
 
 	return 0;
